### PR TITLE
Use SmartManager to store all useful class instances

### PR DESCRIPTION
**Pull request by joshwong382**
Extracted static LinkedList stores from Loopable, SmartDevice and SmartSensorBase into SmartManager

### CODE REVIEW

File: ESP32-PC-Lighting.ino [hunk 1]
```diff
@@ -75,5 +73,5 @@ void setup() {
 }
 
 void loop() {
-  Loopable::loopall();
+  SmartManager::loopall();
```
Comment (rraboy) [inline ESP32-PC-Lighting.ino:76]: use a proper SmartManager instance instead of static: `smartManager->loopall()`

File: SmartManager.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+void SmartManager::loopall() {
+    for (auto it = loopables.begin(); it != loopables.end(); ++it) {
+        if ((*it) != NULL) {
+            (*it)->loop();
+        }
+    }
+}
+
+LinkedList<Loopable*> SmartManager::loopables = LinkedList<Loopable*>([](const Loopable* ptr) {
```
Comment (rraboy) [inline SmartManager.cpp:11]: make this as a field instance(so as devices and sensors) of SmartManager instead of global var

File: SmartManager.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+class SmartManager {
+    public:
+        static void loopall();
+
+    private:
+        // Contributers
+        friend class Loopable;
+        friend class SmartDevice;
+        friend class SmartSensorBase;
```
Comment (rraboy) [inline SmartManager.h:14]: i don't see the point of making these classes `friend` of SmartManager

File: SmartManager.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        // Consumers
+        friend class WebServer;
+
+        static LinkedList<Loopable*> loopables;
+        static LinkedList<SmartDevice*> devices;
+        static LinkedList<SmartSensorBase*> sensors;
+
+        static void addLoopable(Loopable* loopable);
+        static void addDevice(SmartDevice* dev);
+        static void addSensor(SmartSensorBase* dev);
```
Comment (rraboy) [inline SmartManager.h:25]: no statics methods, only instance methods

File: loopable.cpp [hunk 5]
```diff
@@ -1,25 +1,12 @@
 #include "loopable.h"
+#include "SmartManager.h"
 
 Loopable::Loopable() {
-    loopables.add(this);
+    SmartManager::addLoopable(this);
```
Comment (rraboy) [inline loopable.cpp:5]: move the adding of Loopables on the `setup()`

File: loopable.cpp [hunk 6]
```diff
@@ -1,25 +1,12 @@
...
 }
 
 Loopable::Loopable(const bool enable) {
     if (enable) {
-      loopables.add(this);
+      SmartManager::addLoopable(this);
```
Comment (rraboy) [inline loopable.cpp:10]: if you move this in the setup(), then you know right away which one was registered/enabled.

File: smartsensors.cpp [hunk 7]
```diff
@@ -20,7 +15,7 @@ SmartSensorBase::SmartSensorBase(const String _name, const SensorTypes _type) :
 }
 
 void SmartSensorBase::constructor() {
-    allsensors.add(this);
+    SmartManager::addSensor(this);
```
Comment (rraboy) [inline smartsensors.cpp:18]: same, move to setup()

File: smartdevices.cpp [hunk 8]
```diff
@@ -19,7 +14,7 @@ void SmartDevice::updateStatusChanged(const FrontEnd deviceid) {
 void SmartDevice::constructor(const String& _name) {
     name = _name;
     power = PWR_OFF;
-    alldevices.add(this);
+    SmartManager::addDevice(this);
```
Comment (rraboy) [inline smartdevices.cpp:17]: same, move to setup()

File: webserver.cpp [hunk 9]
```diff
@@ -28,7 +29,7 @@ WebServer::~WebServer() {
 
 void WebServer::begin() {
 
-    for (auto it = SmartDevice::alldevices.begin(); it != SmartDevice::alldevices.end(); ++it) {
+    for (auto it = SmartManager::devices.begin(); it != SmartManager::devices.end(); ++it) {
```
Comment (rraboy) [inline webserver.cpp:32]: inject SmartManager's instance into the WebServer and use that instance to iterate thru: `smartManager->devices` better yet, you can provide a method that provide an iterator to your devices

File: loopable.cpp (hunk 5 above)
Comment (joshwong382) [reply loopable.cpp:5]: I'm not sure what you mean. Removing this functionality will remove the whole purpose of Loopable

**Pull request merged**